From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Octavian Purdila <tavip@google.com>
Date: Mon, 26 Oct 2015 15:57:34 +0200
Subject: [PATCH 1/3] Relocate aux entries for NT weak externals

Currently the tagndx for NT weak externals are not relocated when
creating a relocatable object which breaks relocations when doing the
final link of the relocatable object.

For the following example:

    a.c:
        void foo(void);

            int main()
            {
                    foo();
                    return 0;
            }

    b.c:
            #include <stdio.h>

            void __attribute__((weak)) foo(void)
            {
                    printf("%s weak\n", __func__);
            }

$ i686-w64-mingw32-ld -r -o r.o a.o b.o

This is how the symbol table looks for b.o:

$ i686-w64-mingw32-objdump -x b.o | grep -A 1 foo
[ 23](sec  1)(fl 0x00)(ty   0)(scl   2) (nx 0) 0x00000000 .weak._foo.
[ 24](sec  0)(fl 0x00)(ty  20)(scl 105) (nx 1) 0x00000000 _foo

And this is how the symbol table looks for r.o:

$ i686-w64-mingw32-objdump -x r.o | grep -A 1 foo
[ 45](sec  0)(fl 0x00)(ty  20)(scl 105) (nx 1) 0x00000000 _foo
AUX lnno 1 size 0x0 tagndx 23
...
[ 48](sec  1)(fl 0x00)(ty   0)(scl   2) (nx 0) 0x00000000 .weak._foo.

As you can see the AUX entry for _foo is wrong in r.o, it should be 48
but it is 23.

The patch uses the sym_indices array to relocate the AUX entries for
weak externals in _bfd_coff_link_input_bfd.

In order to do so we need the .weak. counterpart symbol to be loaded
before the weak external symbol. This is prevented by the ISFCN()
check _bfd_coff_link_input_bfd and is caused by the fact that the AUX
entry type is 0. To fix this issue we patch gas to propagate the type
of external weak symbols to its .weak. counterpart symbol.

Note that we perform the AUX tag relocations only int the case of
relocatable output because weak external relocation handling
(_bfd_coff_generic_relocate_section) uses the tagndx relative to the
input object.

bfd/

2015-10-26  Octavian Purdila  <octavian.purdila@intel.com>

    * cofflink.c (_bfd_coff_link_input_bfd): relocate AUX entries for
    weak externals when generating relocatable objects

gas/

2015-10-26 Octavian Purdila  <octavian.purdila@intel.com>

    * obj-coff.c (coff_frob_symbol): propagate the type of external
    weak symbol to its .weak. counterpart
---
 bfd/cofflink.c        | 14 +++++++++++---
 gas/config/obj-coff.c |  1 +
 2 files changed, 12 insertions(+), 3 deletions(-)

diff --git a/bfd/cofflink.c b/bfd/cofflink.c
index d029bec..bc7820f 100644
--- a/bfd/cofflink.c
+++ b/bfd/cofflink.c
@@ -2124,7 +2124,8 @@ _bfd_coff_link_input_bfd (struct coff_final_link_info *flaginfo, bfd *input_bfd)
       if ((*indexp < 0
 	   || (bfd_size_type) *indexp < syment_base)
 	  && (*sym_hash == NULL
-	      || (*sym_hash)->auxbfd != input_bfd))
+	      || (*sym_hash)->auxbfd != input_bfd)
+	  && isymp->n_sclass != C_NT_WEAK)
 	esym += add * isymesz;
       else
 	{
@@ -2194,8 +2195,15 @@ _bfd_coff_link_input_bfd (struct coff_final_link_info *flaginfo, bfd *input_bfd)
 		      auxp->x_file.x_n.x_n.x_offset = STRING_SIZE_SIZE + indx;
 		    }
 		}
-	      else if ((isymp->n_sclass != C_STAT || isymp->n_type != T_NULL)
-		       && isymp->n_sclass != C_NT_WEAK)
+	      else if (isymp->n_sclass == C_NT_WEAK)
+	        {
+		    unsigned int *tagndx = &auxp->x_sym.x_tagndx.u32;
+
+		    if (bfd_link_relocatable (flaginfo->info) &&
+			flaginfo->sym_indices[*tagndx] > 0)
+		      *tagndx = flaginfo->sym_indices[*tagndx];
+		}
+	      else if (isymp->n_sclass != C_STAT || isymp->n_type != T_NULL)
 		{
 		  unsigned long indx;

diff --git a/gas/config/obj-coff.c b/gas/config/obj-coff.c
index 7732c0a..5df3940 100644
--- a/gas/config/obj-coff.c
+++ b/gas/config/obj-coff.c
@@ -1259,6 +1259,7 @@ coff_frob_symbol (symbolS *symp, int *punt)
 					   symbol_get_value_expression (weakp));
 	      symbol_set_frag (symp, symbol_get_frag (weakp));
 	      S_SET_SEGMENT (symp, S_GET_SEGMENT (weakp));
+	      S_SET_DATA_TYPE (symp, S_GET_DATA_TYPE (weakp));
 	    }
 	  else
 	    {
